ui/core/player.ts

```diff
@@ -657,6 +674,61 @@ export class Player<SpecType extends Spec> {
 		this.bonusStatsChangeEmitter.emit(eventID);
 	}
 
+	getMeleeCritCapInfo(): MeleeCritCapInfo {
+		const meleeCrit = (this.currentStats.finalStats?.stats[Stat.StatMeleeCrit] || 0.0) / Mechanics.MELEE_CRIT_RATING_PER_CRIT_CHANCE;
+		const meleeHit = (this.currentStats.finalStats?.stats[Stat.StatMeleeHit] || 0.0) / Mechanics.MELEE_HIT_RATING_PER_HIT_CHANCE;
+		const expertise = (this.currentStats.finalStats?.stats[Stat.StatExpertise] || 0.0) / Mechanics.EXPERTISE_PER_QUARTER_PERCENT_REDUCTION / 4;
+		const agility = (this.currentStats.finalStats?.stats[Stat.StatAgility] || 0.0) / this.getClass();
+		const suppression = 4.8;
+		const glancing = 24.0;
+
+		const hasOffhandWeapon = this.getGear().getEquippedItem(ItemSlot.ItemSlotOffHand)?.item.weaponSpeed !== undefined;
+		const meleeHitCap = hasOffhandWeapon ? 27.0 : 8.0;
+		const expertiseCap = this.getInFrontOfTarget() ? 20.5 : 6.5;
+
+		const remainingMeleeHitCap = Math.max(meleeHitCap - meleeHit, 0.0);
+		const remainingExpertiseCap = Math.max(expertiseCap - expertise, 0.0)
+
+		let specSpecificOffset = 0.0;
+
+		if(this.spec === Spec.SpecEnhancementShaman) {
+			// Elemental Devastation uptime is near 100%
+			const ranks = (this as Player<Spec.SpecEnhancementShaman>).getTalents().elementalDevastation;
+			specSpecificOffset = 3.0 * ranks;
+		}
+
+		let debuffCrit = 0.0;
+
+		const debuffs = this.sim.raid.getDebuffs();
+		if (debuffs.totemOfWrath || debuffs.heartOfTheCrusader || debuffs.masterPoisoner) {
+			debuffCrit = 3.0;
```

Comment:
Crit from debuffs has to be manually adjusted here -the display seems to handle adjusting that, and not the player abstraction.  

I've only spent about two hours with this codebase, so let me know if there's a better way to find that total aggregate value.